### ISSUE TITLE
Failed Assert("status == S3PartStatusLoaded")

### ISSUE THREAD

**Issue by homper** [orioledb/orioledb#356]
# Commit
[pre_merge](https://github.com/orioledb/orioledb/commit/2720b49599eae036041eba3085253858e1d31c8d) (2720b49599eae036041eba3085253858e1d31c8d)

# Steps to reproduce
1. create file t/s3_fail_test.py with content:
```py
import logging
from tempfile import mkdtemp

import testgres

from .s3_base_test import S3BaseTest

log = logging.getLogger('werkzeug')
log.setLevel(logging.ERROR)

class S3FailTest(S3BaseTest):
	def test_s3_insert_checkpoint_shutdown(self):
		node = self.node
		node.append_conf(f"""
			orioledb.s3_mode = true
			orioledb.s3_host = '{self.host}:{self.port}/{self.bucket_name}'
			orioledb.s3_region = '{self.region}'
			orioledb.s3_accesskey = '{self.access_key_id}'
			orioledb.s3_secretkey = '{self.secret_access_key}'
			orioledb.s3_cainfo = '{self.ssl_key[0]}'
			orioledb.s3_num_workers = 3

			archive_mode = on
			archive_library = 'orioledb'
		""")
		node.append_conf(f"""
			orioledb.recovery_pool_size = 1
			orioledb.recovery_idx_pool_size = 1
		""")
		node.start()
		node.safe_psql("""
			CREATE EXTENSION orioledb;
			CREATE TABLE o_test_1 (
				val_1 int
			) USING orioledb;
			INSERT INTO o_test_1 SELECT * FROM generate_series(1, 5);
		""")
		node.safe_psql("CHECKPOINT;")
		self.assertEqual([(1, ), (2, ), (3, ), (4, ), (5, )],
		                 node.execute("SELECT * FROM o_test_1"))
		node.safe_psql("""
			INSERT INTO o_test_1 SELECT * FROM generate_series(11, 15);
		""")
		self.assertEqual([(1, ), (2, ), (3, ), (4, ), (5, ), (11, ), (12, ),
		                  (13, ), (14, ), (15, )],
		                 node.execute("SELECT * FROM o_test_1"))
		node.stop()

		new_temp_dir = mkdtemp(prefix=self.myName + '_tgsb_')

		with testgres.get_new_node('test', base_dir=new_temp_dir) as new_node:
			self.loader.download(self.bucket_name, new_node.data_dir)
			new_node.port = self.getBasePort() + 1
			new_node.append_conf(port=new_node.port)

			new_node.start()
			self.assertEqual([(1, ), (2, ), (3, ), (4, ), (5, )],
			                 new_node.execute("SELECT * FROM o_test_1"))
			new_node.stop()
			new_node.cleanup()
```
2. run it as
```sh
make USE_PGXS=1 testgrescheck_part_1 TESTGRESCHECKS_PART_1=" t.s3_test.S3Test.test_s3_insert_checkpoint_shutdown"
```

# Result
postgresql.log:
```
2024-03-18 16:30:27.472 GMT [410265] LOG:  registered custom resource manager "OrioleDB resource manager" with ID 129
2024-03-18 16:30:27.525 GMT [410265] LOG:  OrioleDB initial files scan started
TRAP: failed Assert("status == S3PartStatusLoaded"), File: "src/s3/headers.c", Line: 1137, PID: 410265
/home/vagrant/pgsql/bin/postgres(ExceptionalCondition+0x71)[0x55bf19572121]
/home/vagrant/pgsql/lib/orioledb.so(+0x72b5e)[0x7f36c22c3b5e]
/home/vagrant/pgsql/lib/orioledb.so(+0x71ffa)[0x7f36c22c2ffa]
/home/vagrant/pgsql/lib/orioledb.so(+0x7348d)[0x7f36c22c448d]
/home/vagrant/pgsql/lib/orioledb.so(+0x662fe)[0x7f36c22b72fe]
/home/vagrant/pgsql/bin/postgres(CreateSharedMemoryAndSemaphores+0x144)[0x55bf19407cf4]
/home/vagrant/pgsql/bin/postgres(PostmasterMain+0x886)[0x55bf1939c756]
/home/vagrant/pgsql/bin/postgres(main+0x211)[0x55bf190a3df1]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f36c44aad90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f36c44aae40]
/home/vagrant/pgsql/bin/postgres(_start+0x25)[0x55bf190a4125]
```

More detailed stacktrace which I got on osx:
```
TRAP: failed Assert("status == S3PartStatusLoaded"), File: "src/s3/headers.c", Line: 1137, PID: 97770
0   postgres                            0x000000010087fb2d ExceptionalCondition + 189
1   orioledb.dylib                      0x0000000101703626 initial_parts_counting_callback + 726
2   orioledb.dylib                      0x0000000101702160 iterate_files + 448
3   orioledb.dylib                      0x00000001017009f5 initial_parts_conting + 165
4   orioledb.dylib                      0x00000001017008e4 s3_headers_shmem_init + 420
5   orioledb.dylib                      0x00000001016f06c8 orioledb_shmem_startup + 168
6   postgres                            0x0000000100618fa9 CreateSharedMemoryAndSemaphores + 569
7   postgres                            0x00000001005619d8 PostmasterMain + 3256
8   postgres                            0x000000010040bac2 main + 818
9   dyld                                0x000000010c66152e start + 462
```

# Specs
```
uname -srvmpio
Linux 5.15.0-91-generic #101-Ubuntu SMP Tue Nov 14 13:30:08 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy

pg_config
BINDIR = /home/vagrant/pgsql/bin
DOCDIR = /home/vagrant/pgsql/share/doc
HTMLDIR = /home/vagrant/pgsql/share/doc
INCLUDEDIR = /home/vagrant/pgsql/include
PKGINCLUDEDIR = /home/vagrant/pgsql/include
INCLUDEDIR-SERVER = /home/vagrant/pgsql/include/server
LIBDIR = /home/vagrant/pgsql/lib
PKGLIBDIR = /home/vagrant/pgsql/lib
LOCALEDIR = /home/vagrant/pgsql/share/locale
MANDIR = /home/vagrant/pgsql/share/man
SHAREDIR = /home/vagrant/pgsql/share
SYSCONFDIR = /home/vagrant/pgsql/etc
PGXS = /home/vagrant/pgsql/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE =  '--enable-debug' '--enable-cassert' '--enable-tap-tests' '--with-icu' '--prefix=/home/vagrant/pgsql'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/home/vagrant/pgsql/lib',--enable-new-dtags
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lz -lreadline -lm 
VERSION = PostgreSQL 16.2

gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by homper** [comment]
Closing because it is fixed by small fix in s3 load util (2d0dcd7) and I'll add similar test

**Comment by homper** [comment]
https://github.com/orioledb/orioledb/actions/runs/8368224625/job/22911982692#step:7:327
